### ISSUE TITLE
NOT NULL Constraint for polymorphic_ctype_id?

### ISSUE THREAD

**Issue by guettli** [jazzband/django-polymorphic#148]
Hi,

some how I got rows which had NULL in polymorphic_ctype_id.

AFAIK this makes no sense.

Why not make a non null constraint on this column?

What do you think?

**Comment by vdboor** [comment]
The null requirement is needed, so projects can upgrade to polymorphic, and set the value afterwards.

If the value is null, it likely means you did something like:
* adding polymorphic later.
* imported fixtures
* bypassed the regular save() (or only used save_base())
* removed a ContentType, and set the value to null.

**Comment by TZanke** [comment]
In my opinion, polymorphic should keep my production databases in a consistent state. Allowing Null in the database, breaking the polymorphic plugin, sounds not ideal.

My solution to support both, consistent database and upgrading users would be:

```polymorphic_ctype = models.ForeignKey(ContentType, null=False,...)```
within the polymorphic plugin.

And for users upgrading:

```
class Demo(Polymorphic):
   ...

Demo._meta.get_field('polymorphic_ctype').null = True

```

**Comment by NotSqrt** [comment]
Hi,

I just faced this problem :
When creating instances inside a migration, Django does not give access to the polymorphic manager and other overridable model methods.
That means that ``save()`` won't add the ``polymorphic_ctype`` automatically, and since the DB says that ``NULL`` is acceptable, no integrity exception is raised.

I really want to add this ``not null`` constraint to avoid future similar problems !

Thanks !

**Comment by vdboor** [comment]
good point @NotSqrt I'll reopen this issue so someone can look into this.
I'll accept a pull request that both fixes this issue, and provides the documentation bits to help people migrate to polymorphic. Your migrations will likely fail after all when you add the `polymorphic_ctype_id` as `NOT NULL` - it needs to have a default.

**Comment by bckohan** [comment]
Version 5 will not rely on the contenttypes framework. There is a chance we may still include this for a late version 4.x, but it may also not be worth it...
